feature/sync-measurement-units

FeliGoblin opened the pull request:
## Proposed change:
This is potentially a work in progress, as there's a few ways this can be implemented and I'm not sure what the best way is.. Feedback appreciated 😅

Initially I wanted/envisioned a single entity for each value (such as today_feeding_quantity) which would show the measurement unit chosen in the petlibro account settings, and the value/state converted appropriately. This alone is simple to do (and I can still add it if you like), but it comes with a few drawbacks.
Since PetLibro lets you choose feed units in either weight or volume, and one of those (cups) isn't supported in HA, this means an entity like this can't have a Device Class, which means..
- Users can't choose another unit of measurement from the entity's dropdown
- Long-term statistics will not work, only short-term (10 days)
- If changing from or to mL or cups, you'll get a repair warning for each feed entity like this:
<details><summary>repair warning</summary>

![Screenshot_9](https://github.com/user-attachments/assets/b403d2af-f28e-4c55-a703-0f21d860fc68)

</details>  
It would be more accurate to say those entities can't have a <b><i>static</i></b> Device Class. They <i>can</i> have a dynamic one which would solve the first issue, but the third warning would still happen. I'm not sure how long-term statistics are affected.<br/>
<br/>
<br/>
Currently, I have it so that each value has 2 sensors, one for weight and one for volume..<br/>

![Screenshot_1](https://github.com/user-attachments/assets/e1e88033-09dd-4ebf-b77f-848937bf40da)

<br/>
This allows them to have long-term statistics and the ability to change to another unit supported by HA, which does NOT include cups..<br/>

![Screenshot_10](https://github.com/user-attachments/assets/3157ed35-167f-46cd-a835-99ab7d45116e)

<br/>
If the user has oz or g selected in their account when adding the integration, then the feed weight entities will default to that unit. But that's only when the entity is first created, if they later change to the other weight unit for feed in their account then the entities won't update automatically, they'll have to change their entities manually with the dropdown. This also applies to water or weight value entities, such as the water fountain weight entity.<br/>
<br/>
To change the entities automatically we'd either need to remove their device classes and accept the limitations listed above, OR we <i>could</i> change the dropdown unit choice for them, but that's not really the job of an integration. We could have a checkbox in the account settings form called "Update Entities" or "Sync Entities" or something like that, then the integration can change the entities. It's still not really the job of an integration, but it can be done and having the user confirm it with that checkbox means we're not necessarily betraying the user by fiddling with their front-end user preferences without asking.<br/>
<br/>
These entities also have the 4 PetLibro feed unit options plus a "portion" unit (API "grain") values in the extra attributes..<br/>

![Screenshot_11](https://github.com/user-attachments/assets/ad873ad6-16e1-4ee7-8443-8bfa70018361)

<br/>
As for the number entities, I have 2 of those for each value as well. The first one has no device class and dynamically updates it's values based on the user's feed unit. Although it has no device class it doesn't seem to trigger that repair warning like sensors do. The second entity does have a device class so it should log long-term statistics, and it uses the portion value (1 to 12). There's no "portion" device class or anything like it, so I chose power_factor, which allows a blank unit of measurement. The point of this second entity is just to have something with a device class to log the manual feed amount statistics long-term, it could just be a sensor as well. <br/>

![Screenshot_12](https://github.com/user-attachments/assets/11483026-8292-4b9d-9e99-943b60e89781)
![Screenshot_13](https://github.com/user-attachments/assets/96aa3898-782e-401e-8db3-c6de106ea3d2)

<br/><br/>

![Screenshot_14](https://github.com/user-attachments/assets/27eb3f3b-0a0b-4184-8b0a-3867e3551a60)
![Screenshot_15](https://github.com/user-attachments/assets/035be47b-bc30-4209-b08c-75422c6316ed)


Closes #133 #131 #8 (issue)

## Type of change:

- [ ] New device.
- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature or enhancement (non-breaking change which adds functionality).
- [ ] Documentation only.
- [ ] Other (please explain).

## Checklist:

- [ ] If applicable, I have added corresponding documentation changes.
- [x] If applicable, I have reviewed the [feature / enhancement](https://github.com/jjjonesjr33/petlibro/wiki/Feature-&-Enhancement-Guidelines) guidlines before submitting my request.
- [x] If applicable, I have tested my code for new features & regressions on the latest version of Home Assistant.

## Additional notes:
This is a draft for now, getting this to work exactly how I wanted/envisioned isn't possible, at least not without an overcomplicated process of stuff that integrations generally don't do/aren't used for. Feedback is greatly appreciated :)

C4-Dimitri commented:
I like all of this - but i dont think having 2 number entities makes sense. We dont need to log long term statistics for what the manual feed amount is for example, and it's the job of sensors to report statistics not input entities like the number one, hence why they dont report an error for the unit changing.

I think the way you have implemented everything else is great. The extra attributes ensure users always have access to all of the data should they wish.
I would also say we should implement your update entities button. We could do this in the options flow, but we could also make a button  entity perhaps?

FeliGoblin commented:
> I like all of this - but i dont think having 2 number entities makes sense. We dont need to log long term statistics for what the manual feed amount is for example, and it's the job of sensors to report statistics not input entities like the number one, hence why they dont report the an error for the unit changing.
> 
> I think the way you have implemented everything else is great. The extra attributes ensure users always have access to all of the data should they wish. I would also say we should implement your update entities button. We could do this in the options flow, but we could also make a button entity perhaps?

That makes sense for the number entity, and I completely agree.

For a button entity are you thinking of one per device? Or would it be an entity with no device that updates everything?

C4-Dimitri commented:
Id be thinking one per device. But i guess you are right, the options flow is for the entire integration, so that doesn't really work either.

jjjonesjr33 commented:
> To change the entities automatically we'd either need to remove their device classes and accept the limitations listed above, OR we could change the dropdown unit choice for them, but that's not really the job of an integration. We could have a checkbox in the account settings form called "Update Entities" or "Sync Entities" or something like that, then the integration can change the entities. It's still not really the job of an integration, but it can be done and having the user confirm it with that checkbox means we're not necessarily betraying the user by fiddling with their front-end user preferences without asking.

@FeliGoblin I think the repair warning would work fine—I’ve seen that behavior with other add-ons and plugins before. At least in that case, it does notify the user when something changes.

I like the idea of something like **Sync Entities**. It could either be added to account settings, or even trigger automatically once a user updates their preferences. Do you think you’d be able to make the integration automatically save/reload whenever someone selects a different unit of measurement? The only other integration I know of that handles something like this is the iCloud3 plugin that I use. If need be I can take a look and see how they do it to figure something out for us.

> We dont need to log long term statistics for what the manual feed amount is for example, and it's the job of sensors to report statistics not input entities like the number one, hence why they dont report an error for the unit changing.

@C4-Dimitri — agreed.

Personally, I really value having access to long-term statistics—especially since Petlibro locks that behind their **Care subscription**. That said, if it’s just about tracking the options we set ourselves, that doesn’t really need to be logged. Core tracking, like how much food was dispensed in a day and how many times they ate, is very useful. But something like “What did I set my feeder amount to for that day” feels unnecessary.

All in all though, you’re doing great @FeliGoblin — keep it up, you’re definitely on the right track 😄

FeliGoblin commented:
Alright, I've removed the "portion" number entities. 
I updated the new Dockstream 2 entities to work with this system.
And I added this checkbox to Account Settings.
![Screenshot_20](https://github.com/user-attachments/assets/ff3c6bf0-641f-4b85-a55c-418aaa65bbd2)

By default now, when a user changes a unit of measurement in the options flow, sensors that use that unit type will switch to the chosen unit. Changing your Weight unit will update things like the Fountain "Current Weight" entity. Changing your Feed unit to Grams or Ounces will update things like the Pet Feeder "Last Feed Quantity (Weight)" entity, but not the "Last Feed Quantity (Volume)" entity. Likewise changing to Milliliters will update "Last Feed Quantity (Volume)" but not "Last Feed Quantity (Weight)". Changing to Cups won't update anything since they're not supported in HA. 

If the user selects "Update measurement unit of all Petlibro unit sensor entities" regardless of what unit of measurement is or isn't changed, all sensor entities that use a Feed, Weight or Water measurement unit will be updated.

<details><summary>The main code that handles updating entities inside async_step_account_settings</summary>

``` python

from homeassistant.helpers.entity_registry import async_get as get_entity_registry

if update_setting or update_all_units:
    registry = get_entity_registry(self.hass)
    for unit_type in self.hub.unit_sensor_unique_ids:
        unit = (input if isinstance(input := update_setting.get(unit_type), Unit)
            else Unit(input) if input else getattr(self.member, unit_type, None))
        if (unit_type not in update_setting or not unit or not unit.device_class) and not update_all_units:
            continue
        _LOGGER.debug("Updating %s sensor entities", unit_type)
        if update_all_units and unit_type == API.FEED_UNIT:
            target_units = {"weight": unit if unit.device_class == "weight" else Unit.GRAMS,
                "volume": unit if unit.device_class == "volume" else Unit.MILLILITERS}
        else:
            target_units = {unit.device_class: unit}
        for device_class, target_unit in target_units.items():
            display_precision = ROUNDING_RULES.get(target_unit, 0)
            options = { "unit_of_measurement": target_unit.symbol,
                        "display_precision": display_precision,
                        "suggested_display_precision": display_precision }
            for unique_id in self.hub.unit_sensor_unique_ids.get(unit_type, {}).get(device_class, []):
                entity_id = registry.async_get_entity_id(Platform.SENSOR, DOMAIN, unique_id)
                _LOGGER.debug("Setting %s to %s with display precision %s", entity_id, unit.symbol, display_precision)
                registry.async_update_entity_options(entity_id, Platform.SENSOR, options)
                            
```

</details> 

After that, the user can change the entities to any unit they want if they wish to in the normal way on the front-end.
Thoughts? :)


I've also done all the language translations for this feature as well as the Account Settings feature. I don't know how they're usually done for custom integrations but I had ChatGPT (GPT-5) do them, so I can't say how accurate they are. I believe Lokalise (what HA uses) use GPT-5 as well though.

C4-Dimitri commented:
@FeliGoblin, what do you mean except for Cups? We can't convert to this at all not even for volume?

Not that i care about it, but i feel some users will be quite upset about that if that is the case.

FeliGoblin commented:
> @FeliGoblin, what do you mean except for Cups? We can't convert to this at all not even for volume?
> 
> Not that i care about it, but i feel some users will be quite upset about that if that is the case.

Unfortunately not, there's no cups in UnitOfVolume or SensorDeviceClass.VOLUME so it won't work. It is in the extra attributes though. I did wonder if another entity just for cups might be wanted, it'll mean a third entity per feed-related sensor though 😅

C4-Dimitri commented:
Ah, that makes sense.

Yeah lets stick with the regular volume entity and have it in the extra_attributes instead then.

And if someone cares to have the history for cups specifically, they can make a template sensor that references the extra_attributes instead I guess.

jjjonesjr33 commented:
@FeliGoblin so I found kind of a workaround for cups. It’s a little weird, but hear me out:

Instead of trying to expose cups directly, we can just convert them to something HA already supports (liters/milliliters). For reference, 1 US cup = 236.588 mL ≈ 0.237 L.

```python
# Conversion constant
CUP_TO_ML = 236.588

# Example value in cups
cups_value = 3

# Convert to mL (preferred for HA volume)
ml_value = cups_value * CUP_TO_ML

# Sensor attributes
self._attr_device_class = SensorDeviceClass.VOLUME
self._attr_native_unit_of_measurement = UnitOfVolume.MILLILITERS
self._attr_native_value = ml_value
```
That way HA handles it natively as liters/milliliters, but we can still think in “cups” when we’re working with it.

C4-Dimitri commented:
@jjjonesjr33, @FeliGoblin 
Just another thought, can we just not give a UOM when it's meant to be cups? Then we can say that anytime the sensor has a blank UOM its cups? Is a blank UOM allowed with sensordeviceclass.volume? It will probably throw an error in the logs though, like we recently had for the desiccant_remaining day sensors.

jjjonesjr33 commented:
@C4-Dimitri 

So if you set `device_class=SensorDeviceClass.VOLUME`, HA is gonna expect a valid `UnitOfVolume` (L, mL, gal, ft³, m³). If you leave the UOM blank there, it’ll almost definitely throw warnings or errors in the logs (just like what happened with the `desiccant_remaining` sensors).

If you don’t set a device_class at all (`None`), then you *can* leave the UOM blank (or just put `"cups"`), and HA will treat it like a generic sensor. That won’t cause errors, but you also lose out on the native volume conversions HA normally gives you.

FeliGoblin commented:
> @FeliGoblin so I found kind of a workaround for cups. It’s a little weird, but hear me out:
> 
> Instead of trying to expose cups directly, we can just convert them to something HA already supports (liters/milliliters). For reference, 1 US cup = 236.588 mL ≈ 0.237 L.
> 
> That way HA handles it natively as liters/milliliters, but we can still think in “cups” when we’re working with it.

I'm doing this already I think :)
This is part of the DEVICE_SENSOR_MAP dictionary in sensor.py:
``` python

        PetLibroSensorEntityDescription[GranarySmartCameraFeeder](
            key="today_feeding_quantity_weight",
            translation_key="today_feeding_quantity_weight",
            name="Today Feeding Quantity (Weight)",
            icon="mdi:scale",
            native_unit_of_measurement=UnitOfMass.GRAMS,
            suggested_unit_of_measurement_fn=lambda m: getattr(UnitOfMass, m.feedUnitType.name, None),
            value_fn=lambda d,m: Unit.convert_feed(d.today_feeding_quantity, None, Unit.GRAMS, True),
            device_class=SensorDeviceClass.WEIGHT,
            state_class=SensorStateClass.TOTAL_INCREASING,
            extra_state_attributes_fn=lambda d, m: {"portion": d.today_feeding_quantity}|{
                unit.symbol: Unit.convert_feed(d.today_feeding_quantity, None, unit, True) 
                for unit in VALID_UNIT_TYPES[API.FEED_UNIT] if unit
            },
            petlibro_unit=API.FEED_UNIT
        ),
        PetLibroSensorEntityDescription[GranarySmartCameraFeeder](
            key="today_feeding_quantity_volume",
            translation_key="today_feeding_quantity_volume",
            name="Today Feeding Quantity (Volume)",
            icon="mdi:scale",
            native_unit_of_measurement=UnitOfVolume.MILLILITERS,
            value_fn=lambda d,m: Unit.convert_feed(d.today_feeding_quantity, None, Unit.MILLILITERS, True),
            device_class=SensorDeviceClass.VOLUME,
            state_class=SensorStateClass.TOTAL_INCREASING,
            extra_state_attributes_fn=lambda d, m: {"portion": d.today_feeding_quantity}|{
                unit.symbol: Unit.convert_feed(d.today_feeding_quantity, None, unit, True) 
                for unit in VALID_UNIT_TYPES[API.FEED_UNIT] if unit
            },
            petlibro_unit=API.FEED_UNIT
        ),

```

That Unit.convert_feed() function takes None as a unit argument to represent the raw API or portion value (12th of a cup)

jjjonesjr33 commented:
@FeliGoblin, @C4-Dimitri — for the `Manual Feed Quantity` setting, do you think it would make sense to switch from a slider to a dropdown with selectable portions now that we’re increasing the portion options? What do you think?

Linking https://github.com/jjjonesjr33/petlibro/issues/141

C4-Dimitri commented:
@jjjonesjr33, That might work well for things like cups, but not total grams (imo). So i would say no.

jjjonesjr33 commented:
> @jjjonesjr33, That might work well for things like cups, but not total grams (imo). So i would say no.

Gotcha, that makes sense.

jjjonesjr33 reviewed (approved):
Looks good @FeliGoblin, thanks for your update!

jjjonesjr33 commented:
@FeliGoblin Will have to recreate the pull request, there was an issue with the account settings as well problems with the unit conversion. 

#### Other problems noticed
- Fountains we listed as `lb`
